### PR TITLE
Wip/extensibility

### LINKED ISSUE #5: Add hreflang attribute to sitemap url entries
For multilingual sites adding the `hreflang` attribute is required for correct SEO.

I would suggest that the `locale` property of a menu item should be used for that. If not present no `hreflang` attribute should be attached.

Problems to solve:

Automaticly add locale versions of static pages if the translate plugin is present and has more then one active locale.

I could provide a PR for that.

### PR DESCRIPTION
**Pull request by der-On**
This pull request adds functionality to extend sitemap elements either by blocking theire creation or by extending them.

closes #5

### CODE REVIEW

**Comment by mjauvin** [comment]
Can you explain how these events can be used to make a full multi-lang sitemap? You need to be able to inject at a much earlier stage and you'll need to generate the page url for each locale.

File: models/Definition.php [hunk 1]
```diff
@@ -189,12 +190,15 @@ protected function addItemToSet($item, $url, $mtime = null)
...
+
+
         $urlElement = $this->makeUrlElement(
             $xml,
             $url,
             $mtime,
             $item->changefreq,
-            $item->priority
+            $item->priority,
+            $item
```
[the remark this answers is not in the document]
Comment (der-On) [reply models/Definition.php:201]: true

**Comment by der-On** [comment]
> Can you explain how these events can be used to make a full multi-lang sitemap? You need to be able to inject at a much earlier stage and you'll need to generate the page url for each locale.

For my use case it worked as i only needed to add xhtml:link nodes and attach a query string to each url, but one might actually need to pass additional data to the item to e.g. reference the original resource it was generated from. I will further investiage and adjust this PR.

**Comment by der-On** [comment]
@mjauvin I've removed the unessecary arguments and will be trying to extend the translate plugin to allow for localized page urls in the sitemap.

**Comment by der-On** [comment]
@mjauvin @LukeTowers I've experimented with the translate plugin to add localized links to the sitemap and came up with this code:

```php
// add localized links to static pages in sitemap
Event::listen('winter.sitemap.makeUrlElement', function($definition, $xml, $pageUrl, $lastModified, $item, $url) {
    $baseUrl = rtrim(url('/'), '/');

    if ($item->type === 'all-static-pages' || $item->type === 'static-page') {
        $locales = array_keys(Locale::listAvailable());

        foreach($locales as $locale) {
            $localeUrl = $baseUrl . '/' . $locale . str_replace($baseUrl, '', $pageUrl);
            $linkElement = $xml->createElement('xhtml:link');
            $linkElement->setAttribute('rel', 'alternate');
            $linkElement->setAttribute('hreflang', $locale);
            $linkElement->setAttribute('href', $localeUrl);
            $url->appendChild($linkElement);
        }
    }
});
```

**Comment by mjauvin** [comment]
you can't build the url like this, it won't work in all scenarios. please check how I did it in my plugin.

**Comment by mjauvin** [comment]
What about cms pages with a translatable slug?
ref. https://github.com/mjauvin/wn-mlsitemap-plugin/blob/master/classes/Definition.php#L69

**Comment by mjauvin** [comment]
> @mjauvin @LukeTowers I've experimented with the translate plugin to add localized links to the sitemap and came up with this code:
> 
> ```
> // add localized links to static pages in sitemap
> Event::listen('winter.sitemap.makeUrlElement', function($definition, $xml, $pageUrl, $lastModified, $item, $url) {
>     $baseUrl = rtrim(url('/'), '/');
> 
>     if ($item->type === 'all-static-pages' || $item->type === 'static-page') {
>         $locales = array_keys(Locale::listAvailable());
> 
>         foreach($locales as $locale) {
>             $localeUrl = $baseUrl . '/' . $locale . str_replace($baseUrl, '', $pageUrl);
>             $linkElement = $xml->createElement('xhtml:link');
>             $linkElement->setAttribute('rel', 'alternate');
>             $linkElement->setAttribute('hreflang', $locale);
>             $linkElement->setAttribute('href', $localeUrl);
>             $url->appendChild($linkElement);
>         }
>     }
> });
> ```

where is the $url parameter to the event handler coming from? it doesn't match the event signature

**Comment by mjauvin** [comment]
never mind, I was looking at the beforeMakeUrlElement event...

File: README.md [hunk 2]
```diff
@@ -82,3 +82,25 @@ Array (
...
+});
+```
+
+##### Extend created sitemap elements
+
+To extend a created sitemap element do:
+
+```php
+Event::listen('winter.sitemap.makeUrlElement', function ((Definition) $definition, (DomDocument) $xml, (string) $pageUrl, (string) $lastModified, (DefinitionItem) $item, (ElementNode) $urlElement) {
+    $url->appendChild($xml->createElement('bestcmsever', 'WinterCMS');
```
Comment (mjauvin) [inline README.md:104]:
```suggestion
    $urlElement->appendChild($xml->createElement('bestcmsever', 'WinterCMS');
```

File: models/Definition.php [hunk 3]
```diff
@@ -204,19 +206,57 @@ protected function addItemToSet($item, $url, $mtime = null)
...
+        $url->appendChild($xml->createElement('priority', $item->priority));
+
+        /**
+         * @event winter.sitemap.makeUrlElement
+         * Provides an opportunity to interact with a sitemap element after it has been generated.
+         *
+         * Example usage:
+         *
+         *     Event::listen('winter.sitemap.makeUrlElement', function ((Definition) $definition, (DomDocument) $xml, (string) $pageUrl, (string) $lastModified, (DefinitionItem) $item, (ElementNode) $urlElement) {
+         *         $url->appendChild($xml->createElement('bestcmsever', 'OctoberCMS');
```
Comment (mjauvin) [inline models/Definition.php:255]:
```suggestion
         *         $urlElement->appendChild($xml->createElement('bestcmsever', 'OctoberCMS');
```

File: models/Definition.php [hunk 4]
```diff
@@ -204,19 +206,57 @@ protected function addItemToSet($item, $url, $mtime = null)
...
+
         $this->urlCount++;
 
         $url = $xml->createElement('url');
         $url->appendChild($xml->createElement('loc', $pageUrl));
         $url->appendChild($xml->createElement('lastmod', $lastModified));
-        $url->appendChild($xml->createElement('changefreq', $frequency));
-        $url->appendChild($xml->createElement('priority', $priority));
+        $url->appendChild($xml->createElement('changefreq', $item->changefreq));
+        $url->appendChild($xml->createElement('priority', $item->priority));
```
Comment (mjauvin) [inline models/Definition.php:246]: It would make sense to rename the `$url` variable to `$urlElement` (less confusion)

**Comment by der-On** [comment]
> Another problem I see is that the URLs needs to be generated for all locales (not just have `xhtml:link` for each locale)

I'm not sure what is the best practise approach here. My quick research told me that using a canonical URL + the alternate URLs for the locales seems to be what google recommends nowdays.

File: models/Definition.php (hunk 4 above)
Comment (der-On) [reply models/Definition.php:246]: I agree. Will do so.

**Comment by der-On** [comment]
> you can't build the url like this, it won't work in all scenarios. please check how I did it in my plugin.

Just checked it, you basically pass localized URLs to the menu item. That could be a way. I would like to also pass a reference (like an ID or code) of the menu item down to the sitemap item, so one has a canonical reference to the original item when handling the events. Maybe merging both could be the way to go. This gives us localized URLs without the need of handling events but while creating menu items and it gives us a reference to the original item when handling the sitemap events.

**Comment by mjauvin** [comment]
> > Another problem I see is that the URLs needs to be generated for all locales (not just have `xhtml:link` for each locale)
> 
> I'm not sure what is the best practise approach here. My quick research told me that using a canonical URL + the alternate URLs for the locales seems to be what google recommends nowdays.

According to this document, it is a bit more complex than what you describe:

https://developers.google.com/search/docs/advanced/crawling/localized-versions#sitemap

**Comment by der-On** [comment]
@mjauvin I'm now passing the item reference to the events, this way the origin of the item can be retrieved within the event handler. I was also able to add localized urls in a sane way in the translate plugin:

```php
$self = $this;

// add localized links to static pages in sitemap
Event::listen('winter.sitemap.makeUrlElement', function($definition, $xml, $pageUrl, $lastModified, $item, $itemReference, $urlElement) use ($self) {
            if ($item->type === 'all-static-pages' || $item->type === 'static-page') {
                if (!$itemReference) {
                    return;
                }

                $locales = array_keys(Locale::listAvailable());

                foreach($locales as $locale) {
                    $linkElement = $xml->createElement('xhtml:link');
                    $linkElement->setAttribute('rel', 'alternate');
                    $linkElement->setAttribute('hreflang', $locale);
                    $linkElement->setAttribute('href', $self->localeUrl($pageUrl, $locale));
                    $urlElement->appendChild($linkElement);
                }
            }
        });
```

**Comment by mjauvin** [comment]
@der-On if your PR allows creating the equivalent code linked below in the translate plugin, I'll approve and merge it:

https://github.com/mjauvin/wn-mlsitemap-plugin/blob/master/classes/Definition.php#L69-L97

File: models/Definition.php [hunk 5]
```diff
@@ -114,7 +114,7 @@ public function generateSitemap()
                      * Single item
                      */
                     if (isset($itemInfo['url'])) {
-                        $this->addItemToSet($item, $itemInfo['url'], array_get($itemInfo, 'mtime'));
+                        $this->addItemToSet($item, $itemInfo['url'], array_get($itemInfo, 'mtime'), array_get($item, 'reference'));
```
Comment (mjauvin) [inline models/Definition.php:117]:
```suggestion
                        $this->addItemToSet($item, $itemInfo['url'], array_get($itemInfo, 'mtime'), $item->reference);
```

File: models/Definition.php [hunk 6]
```diff
@@ -128,7 +128,7 @@ public function generateSitemap()
                         {
                             foreach ($items as $item) {
                                 if (isset($item['url'])) {
-                                    $this->addItemToSet($parentItem, $item['url'], array_get($item, 'mtime'));
+                                    $this->addItemToSet($parentItem, $item['url'], array_get($item, 'mtime'), array_get($item, 'reference'));
```
Comment (mjauvin) [inline models/Definition.php:131]:
```suggestion
                                    $this->addItemToSet($parentItem, $item['url'], array_get($item, 'mtime'), $parentItem->reference);
```

**Review by mjauvin — CHANGES_REQUESTED** [review]
Please fix those issues.

**Comment by mjauvin** [comment]
See #10 for other plugins implementing Multilang functionality.

**Comment by LukeTowers** [comment]
Replaced by #11

**Pull request closed without merging**